### ISSUE TITLE
Broadcasting learning rate when using Horovod with Keras ReduceLROnPlateau callback 

### ISSUE THREAD

**Issue by vbvg2008** [horovod/horovod#1095]
Hi I am using horovod + tf.keras for my multi-gpu training.  I used 4 GPUs for training and 1 GPU(rank0) for validation. Since only rank0 has access to validation loss, I only applied the ReduceLROnPlateau callback on rank0. 

Here's my question:

When lr is decreased on rank0,   will the lr be automatically broadcasted to other ranks?  or do I need to manually broadcast the lr to other ranks?  if latter, how do i do it?

**Comment by alsrgv** [comment]
@vbvg2008, if you add MetricAverageCallback before ReduceLROnPlateau (like [this](https://github.com/horovod/horovod/blob/master/examples/keras_mnist_advanced.py#L91)), all ranks will observe the same validation loss and behave in the same way.  That's the preferred way to accomplish synchronicity with logic based on validation losses.

**Comment by vbvg2008** [comment]
@alsrgv In my case, all my validation data is in form of tfrecords, in order to avoid the reading conflict between ranks, I have to use tf.data.dataset.shard to assign different portion of validation data to each ranks. if my validation load is not balanced (say rank0 has 100 example, other ranks has 10 example), then `MetricAverageCallback` will not produce correct validation metric over all data due to imbalanced validation loads.

In other words, `MetricAverageCallback` will only work correctly when all ranks are given the same validation data or when validation loads are perfectly balanced.   is it right?

**Comment by alsrgv** [comment]
@vbvg2008, makes sense - if you have a large imbalance between workers, the average of averages may not have the same result as a global average.

One approach could be to split your data to a large number of equally-sized tfrecord files, e.g. 100 files per rank.  Then it's very unlikely that imbalance will have a noticeable effect.

Another approach would be to run all validation on rank 0 (if it's not too large), and implement a callback that broadcasts validation metrics from rank 0 to everyone else.

The third approach would be to have the `MetricAverageCallback` re-weight values based on the actual number of the validation steps.  It looks like [this code in Keras](https://github.com/keras-team/keras/blob/master/keras/engine/training_arrays.py#L487) and [this code in Keras](https://github.com/keras-team/keras/blob/a1397169ddf8595736c01fcea084c8e34e1a3884/keras/callbacks.py#L74) would allow us to capture the size of every validation batch (by overriding `on_test_batch_end`), and use it to do the global averaging.

Is the third option something you'd be able to try and contribute?

**Comment by vbvg2008** [comment]
@alsrgv glad to contribute, what do you envision the new use case of `MetricAverageCallback` to be? 
Do you think it's good idea to make `MetricAverageCallback` always be a weighted-averaging based on number of batches?  

Furthermore,  What should the user do if there are some metrics that are not intended to be averaged? (For example, the 'size' metric that you referenced? )

**Comment by alsrgv** [comment]
We could add a flag to `MetricAverageCallback` to perform weighted average (false by default), and also add a metric mapping where you could specify what to do with every metric (default is average) - average, sum, broadcast.  WDYT?

cc @tgaddair

**Comment by vbvg2008** [comment]
@alsrgv In my opinion, if we are going to add `sum` and `broadcast` operations , I would first rename `MetricAverageCallback` to `MetricReduceCallback` because it is essentially doing reduction operations among different processes instead of purely averaging. 

The new callback should have the following 4 operations: `avg`(default), `weighted_avg`, `sum` and `broadcast` . 

I would design the use case to be like this:
```
# Only need to define the metrics that are not meant to be averaged, 
# the rest of the metrics will be averaged by default
metrics_op_map = {"size": "sum",
                  "val_loss": "weighted_avg",
                  "some_custom_metrics": "broadcast"}

callbacks = [...
             hvd.callbacks.MetricReduceCallback(metrics_op_map),
             ...]

```
The only problem though, is `broadcast`  requires the source rank as additional argument, however, designing one more argument only for the broadcast seems a bit unnecessary.  Of course we may assume the source will always 0 :smile: 

Thoughts?

**Comment by alsrgv** [comment]
I like this proposal.  We'd still have to keep `MetricAverageCallback`, but it can delegate to `MetricReduceCallback` with default parameters.

I don't see a big issue with adding an optional `broadcast_rank=0` parameter to `MetricReduceCallback` (assume the source of truth will be the same for all metrics).

We should have another flag which would indicate whether the metric reduction should be done on every step (could be slower), or only at the end of epoch (default).  This will help with use case described in #1092.

Finally, I think it'd be better to use enum (like hvd.Compression) for metric actions.
